### PR TITLE
Support custom toleration on ds

### PR DESCRIPTION
**Pull request by mshitrit**
This PR meant to solve ECOPROJECT-1235, it would allow the user to use SNR configuration in order to create customized tolerations on SNR's daemonset.

### CODE REVIEW

File: controllers/selfnoderemediationconfig_controller.go
Comment (razo7) [inline controllers/selfnoderemediationconfig_controller.go:204]: Can tolerations be nil since this field is [`omitempty`](https://github.com/medik8s/self-node-remediation/pull/130/files#diff-792340dfdee839c7939f417270ed1aa40b9f59b0e2d757b5d5417f7a59fbeb9fR118)? If so , then we should check it first.

Comment (mshitrit) [reply controllers/selfnoderemediationconfig_controller.go:204]: this check covers nil as well.
i.e:
```
len(nil) is 0
```

File: controllers/selfnoderemediationconfig_controller.go
Comment (slintes) [inline controllers/selfnoderemediationconfig_controller.go:209]: - why iterating the tolerations, and not adding all at once?
- just to protect our future selfs from surprises in case we add manifests to `/install`: we know that there is only 1 object today, the daemonset. What about ensuring that it is really 1 object only, and process it directly without iteration? And fail if there are more manifests, which means this code needs an update?

File: controllers/selfnoderemediationconfig_controller.go
Comment (slintes) [inline controllers/selfnoderemediationconfig_controller.go:232]: - `configTolerations` isn't a very descriptive name, is it? 🤔 
- this conversion can be done easier: `newToleration, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&toleration)`

**Review by slintes — CHANGES_REQUESTED** [review]

File: controllers/selfnoderemediationconfig_controller.go
Comment (mshitrit) [reply controllers/selfnoderemediationconfig_controller.go:209]: >why iterating the tolerations, and not adding all at once?

IIUC each toleration needs to be converted from `v1.Toleration` to `map[string]interface{}`

>just to protect our future selfs from surprises in case we add manifests to /install: we know that there is only 1 object today, the daemonset. What about ensuring that it is really 1 object only, and process it directly without iteration? And fail if there are more manifests, which means this code needs an update?

Good idea :+1:

File: controllers/selfnoderemediationconfig_controller.go
Comment (mshitrit) [reply controllers/selfnoderemediationconfig_controller.go:142]: I agree, webhook is probably the best solution here

File: controllers/selfnoderemediationconfig_controller.go
Comment (slintes) [reply controllers/selfnoderemediationconfig_controller.go:209]: > is there a simple way to convert the whole slice ?

probably not, but you can have that loop where you do that conversion IMHO

Comment (mshitrit) [reply controllers/selfnoderemediationconfig_controller.go:209]: If it's not something you feel strongly about I think I prefer to keep as is.
Since we need to iterate over the tolerations in any case I do prefer the pattern of having one method that handles the whole slice which calls a method that handle a single element (toleration).
IMO it's improves readability and is more modular.

File: controllers/selfnoderemediationconfig_controller.go
Comment (mshitrit) [reply controllers/selfnoderemediationconfig_controller.go:204]: Not sure about `can't` but it definitely `shouldn't` be empty

File: controllers/selfnoderemediationconfig_controller.go
Comment (slintes) [reply controllers/selfnoderemediationconfig_controller.go]: It doesn't make a difference on performance in this case, but what kind of pattern is it where you do more work than needed inside the loop? (processing the whole DS object vs just converting the tolerations to unstructured) 🤔 

> IMO it's improves readability

let's not start arguing about this 😎 

>  and is more modular.

if you want to have this more modular, put the conversion of the toleration slice into its own function?

File: api/v1alpha1/selfnoderemediationconfig_webhook.go
Comment (slintes) [inline api/v1alpha1/selfnoderemediationconfig_webhook.go:99]: Better run all validations ad aggregate errors, so users know about all problems in one run.
See e.g. https://github.com/medik8s/node-healthcheck-operator/blob/main/api/v1alpha1/nodehealthcheck_webhook.go#L116

File: api/v1alpha1/selfnoderemediationconfig_webhook.go
Comment (slintes) [inline api/v1alpha1/selfnoderemediationconfig_webhook.go:160]: redundant, the range assignment in the loop can deal with nil and empty slices

File: api/v1alpha1/selfnoderemediationconfig_webhook.go
Comment (slintes) [inline api/v1alpha1/selfnoderemediationconfig_webhook.go:175]: please add: when operator is Exists, value must be empty

File: api/v1alpha1/selfnoderemediationconfig_webhook.go
Comment (slintes) [inline api/v1alpha1/selfnoderemediationconfig_webhook.go:177]: typo: tolerarion (not here only)

File: api/v1alpha1/selfnoderemediationconfig_webhook_test.go
Comment (slintes) [inline api/v1alpha1/selfnoderemediationconfig_webhook_test.go:158]: this test should check if  the error contains all problems

**Review by slintes — CHANGES_REQUESTED** [review]

File: controllers/selfnoderemediationconfig_controller.go
Comment (mshitrit) [reply controllers/selfnoderemediationconfig_controller.go]: > but what kind of pattern is it where you do more work than needed inside the loop? 

Fair point

>if you want to have this more modular, put the conversion of the toleration slice into its own function?

Sure, that WFM :+1:

File: api/v1alpha1/selfnoderemediationconfig_webhook.go
Comment (mshitrit) [reply api/v1alpha1/selfnoderemediationconfig_webhook.go:160]: cool, I didn't know range is valid over nil :+1:

File: api/v1alpha1/selfnoderemediationconfig_webhook_test.go
Comment (mshitrit) [reply api/v1alpha1/selfnoderemediationconfig_webhook_test.go:158]: Hmm maybe I'm missing something IIUC these are valid values, the test check that there are no errors

Comment (slintes) [reply api/v1alpha1/selfnoderemediationconfig_webhook_test.go:173]: oh sorry, jumping too much around in the code. But maybe add it to testMultipleInvalidFields() above?

**Comment by slintes** [comment]
/lgtm

left one comment about testing multiple validation errors, will leave it to you if you want to add that, feel free to unhold :) 
/hold

File: api/v1alpha1/selfnoderemediationconfig_webhook_test.go
Comment (mshitrit) [reply api/v1alpha1/selfnoderemediationconfig_webhook_test.go:176]: Sure, sorry I've initial missed your latest respons -, for some reason this thread didn't automatically update in main conversation tab (I saw it once I switched to files tab).

Thanks for mentioning it again with the lgtm comment :+1:

**Review by slintes — APPROVED** [review]

**Pull request merged**
